src/data_source/fetching.rs

```diff
@@ -340,22 +365,66 @@ where
     Types: NodeType,
 {
     storage: S,
+    height: u64,
     block_notifier: Notifier<BlockQueryData<Types>>,
     leaf_notifier: Notifier<LeafQueryData<Types>>,
 }
 
+impl<Types, S> NotifyStorage<Types, S>
+where
+    Types: NodeType,
+    S: UpdateAvailabilityData<Types>,
+{
+    async fn insert_leaf(&mut self, leaf: LeafQueryData<Types>) -> Result<(), S::Error> {
+        self.leaf_notifier.notify(&leaf);
+        if leaf.height() + 1 > self.height {
+            self.height = leaf.height() + 1;
+        }
+        self.storage.insert_leaf(leaf).await
+    }
+
+    async fn insert_block(&mut self, block: BlockQueryData<Types>) -> Result<(), S::Error> {
+        self.block_notifier.notify(&block);
+        if block.height() + 1 > self.height {
+            self.height = block.height() + 1;
+        }
+        self.storage.insert_block(block).await
+    }
+}
+
+impl<Types, S> NotifyStorage<Types, S>
+where
+    Types: NodeType,
+    S: VersionedDataSource,
+{
+    async fn commit(&mut self) -> Result<(), S::Error> {
+        self.storage.commit().await
+    }
+}
+
 impl<Types, S, P> Fetcher<Types, S, P>
 where
     Types: NodeType,
+    S: NodeDataSource<Types>,
 {
-    fn new(storage: S, provider: P) -> Self {
+    async fn new(storage: S, provider: P) -> Self {
+        // Get the height from storage if possible. If not, it's fine: we'll update this as soon as
+        // we see a new block or leaf anyways.
+        let height = storage.block_height().await.unwrap_or_else(|err| {
```

Comment:
Could this fail in a recoverable way? Just wondering if a panic is more appropriate if failing here means that storage is botched

Reply:
Do nothing and returning 0 _is_ recoverable: it is never _incorrect_ to have a height less than the true height -- this can happen for many reasons including our attached HotShot node lagging, and we will update height as soon as our HotShot client updates.

As for whether storage is botched, it could be, but it could also be a transient failure that will resolve itself, like we temporarily lost connectivity to the DB server.

Reply:
gotcha makes sense